### PR TITLE
match authrealm ldap url to Openshift's ldap url 

### PR DESCRIPTION
**Pull request by leena-jawale**
issue: https://github.com/stolostron/backlog/issues/22283

/hold

### CODE REVIEW

File: controllers/authrealm/dex.go [hunk 1]
```diff
@@ -396,12 +397,26 @@ func (r *AuthRealmReconciler) createDexConnectors(authRealm *identitatemv1alpha1
...
+			var scope string
+			if url.Scope == ldaputil.ScopeSingleLevel {
+				scope = "one"
+			}
+			if url.Scope == ldaputil.ScopeWholeSubtree {
+				scope = "sub"
+			}
+			if url.Scope == ldaputil.ScopeBaseObject {
+				scope = "base"
+			}
```
Comment (itdove) [inline controllers/authrealm/dex.go:413]: could you use a switch statement here ?

File: controllers/authrealm/dex.go [hunk 2]
```diff
@@ -415,16 +430,23 @@ func (r *AuthRealmReconciler) createDexConnectors(authRealm *identitatemv1alpha1
...
 					},
 					GroupSearch: authRealm.Spec.LDAPExtraConfigs[idp.Name].GroupSearch,
 				},
 			}
+			if url.Scheme == "ldap" {
+				c.LDAP.StartTLS = true
+			}
+			if url.Scheme == "ldaps" {
+				c.LDAP.StartTLS = false
+			}
```
Comment (itdove) [inline controllers/authrealm/dex.go:449]: same here (switch) ?

File: controllers/authrealm/dex.go [hunk 3]
```diff
@@ -415,16 +430,23 @@ func (r *AuthRealmReconciler) createDexConnectors(authRealm *identitatemv1alpha1
...
+						Scope:     scope,
+						Username:  url.QueryAttribute,
 						IDAttr:    idp.LDAP.Attributes.ID[0],
 						EmailAttr: idp.LDAP.Attributes.Email[0],
 						NameAttr:  idp.LDAP.Attributes.Name[0],
 					},
 					GroupSearch: authRealm.Spec.LDAPExtraConfigs[idp.Name].GroupSearch,
 				},
 			}
+			if url.Scheme == "ldap" {
```
Comment (chrisahl) [inline controllers/authrealm/dex.go:444]: Is this logic correct or backwards?  Wouldn't ldaps want TLS?

File: go.mod [hunk 4]
```diff
@@ -11,6 +11,7 @@ require (
 	github.com/onsi/gomega v1.18.0
 	github.com/openshift/api v3.9.1-0.20191111211345-a27ff30ebf09+incompatible
 	github.com/openshift/generic-admission-server v1.14.1-0.20210422140326-da96454c926d
+	github.com/openshift/library-go v0.0.0-20210916194400-ae21aab32431
```
Comment (chrisahl) [inline go.mod:14]: Assume we need to wait for idp-client-api change to get merged so you can pull it in with a change to go mod before merging this PR

Comment (leena-jawale) [reply go.mod:14]: yeah thats right

File: controllers/authrealm/dex.go (hunk 3 above)
Comment (leena-jawale) [reply controllers/authrealm/dex.go]: Its other way round. if there ldaps its using TLS connection to server. If there ldap, then wwe have to providee startTLS =true, to upgrade the connection to TLS

Comment (leena-jawale) [reply controllers/authrealm/dex.go]: Screen Shot 2022-05-18 at 8 25 17 AM

File: controllers/authrealm/dex.go [hunk 5]
```diff
@@ -396,12 +397,26 @@ func (r *AuthRealmReconciler) createDexConnectors(authRealm *identitatemv1alpha1
...
+				return nil, fmt.Errorf("Error parsing LDAP URL: %v", err)
+			}
+			var scope string
+			switch url.Scope {
+			case ldaputil.ScopeSingleLevel:
+				scope = "one"
+			case ldaputil.ScopeWholeSubtree:
+				scope = "sub"
+			case ldaputil.ScopeBaseObject:
+				scope = "base"
```
Comment (robinbobbitt) [inline controllers/authrealm/dex.go:411]: The OpenShift docs say they only support "one" or "sub", as does Dex. See https://github.com/dexidp/dex/blob/9d3471e39b24c11350b8c911fc6e736f0be7c785/connector/ldap/ldap.go#L174

What does OpenShift do if you put a scope of base? Seems we should not allow it.

File: controllers/authrealm/dex.go [hunk 6]
```diff
@@ -415,16 +430,24 @@ func (r *AuthRealmReconciler) createDexConnectors(authRealm *identitatemv1alpha1
 						Namespace: authRealm.Namespace,
 					},
 					UserSearch: dexoperatorv1alpha1.UserSearchSpec{
-						BaseDN:    authRealm.Spec.LDAPExtraConfigs[idp.Name].BaseDN,
-						Filter:    authRealm.Spec.LDAPExtraConfigs[idp.Name].Filter,
-						Username:  idp.LDAP.Attributes.PreferredUsername[0],
+						BaseDN:    url.BaseDN,
```
Comment (robinbobbitt) [inline controllers/authrealm/dex.go:433]: Let's update the unit tests to be sure we're confirming these values are being mapped from the URL.

**Review by robinbobbitt — CHANGES_REQUESTED** [review]
Great improvements! I've added a few comments and requests. Thanks!

File: controllers/authrealm/dex.go (hunk 5 above)
Comment (leena-jawale) [reply controllers/authrealm/dex.go:411]: I'll remove base.

File: controllers/authrealm/dex.go (hunk 6 above)
Comment (leena-jawale) [reply controllers/authrealm/dex.go:431]: updated

File: controllers/authrealm/dex.go [hunk 7]
```diff
@@ -396,12 +397,26 @@ func (r *AuthRealmReconciler) createDexConnectors(authRealm *identitatemv1alpha1
 			r.Log.Info("generated intermediate connectors", "cs", cs)
 		case openshiftconfigv1.IdentityProviderTypeLDAP:
 			r.Log.Info("create connector for LDAP")
+			url, err := ldaputil.ParseURL(idp.LDAP.URL)
+			if err != nil {
+				return nil, fmt.Errorf("Error parsing LDAP URL: %v", err)
```
[the remark this answers is not in the document]
Comment (leena-jawale) [reply controllers/authrealm/dex.go:402]: added webhook check in same PR

File: controllers/authrealm/dex.go [hunk 8]
```diff
@@ -399,35 +400,65 @@ func (r *AuthRealmReconciler) createDexConnectors(authRealm *identitatemv1alpha1
...
 						IDAttr:    idp.LDAP.Attributes.ID[0],
 						EmailAttr: idp.LDAP.Attributes.Email[0],
 						NameAttr:  idp.LDAP.Attributes.Name[0],
 					},
 					GroupSearch: authRealm.Spec.LDAPExtraConfigs[idp.Name].GroupSearch,
 				},
 			}
+			switch url.Scheme {
+			case "ldap":
+				if idp.LDAP.Insecure {
```
Comment (robinbobbitt) [inline controllers/authrealm/dex.go:447]: This configuration of insecure connection, not upgraded leaks passwords, and per Dex docs may not be supported in the future. Shall we perhaps not support this option? We could update our webhook to never allow `insecure: true` with our operator.

File: controllers/authrealm/suite_test.go [hunk 9]
```diff
@@ -631,16 +632,187 @@ var _ = Describe("Process AuthRealm LDAP: ", func() {
...
+							IdentityProviderConfig: openshiftconfigv1.IdentityProviderConfig{
+								Type: openshiftconfigv1.IdentityProviderTypeLDAP,
+								LDAP: &openshiftconfigv1.LDAPIdentityProvider{
+									URL:    "ldaps://myldap.example.com:636/dc=example,dc=com?mail,cn?one?(objectClass=person)",
+									BindDN: BindDN,
+									Attributes: openshiftconfigv1.LDAPAttributeMapping{
+										ID:                []string{IDAttr},
+										PreferredUsername: []string{"mail"},
+										Name:              []string{NameAttr},
+										Email:             []string{"mail"},
```
Comment (robinbobbitt) [inline controllers/authrealm/suite_test.go:733]: Better to use `EmailAttr` here?

File: controllers/authrealm/suite_test.go [hunk 10]
```diff
@@ -558,12 +561,12 @@ var _ = Describe("Process AuthRealm LDAP: ", func() {
 							IdentityProviderConfig: openshiftconfigv1.IdentityProviderConfig{
 								Type: openshiftconfigv1.IdentityProviderTypeLDAP,
 								LDAP: &openshiftconfigv1.LDAPIdentityProvider{
-									URL:    "myldap.example.com",
+									URL:    "ldap://myldap.example.com:389/dc=example,dc=com?mail,cn?one?(objectClass=person)",
```
Comment (robinbobbitt) [inline controllers/authrealm/suite_test.go:564]: Shall we build this URL from Filter, BaseDN etc to be more clear how these all tie together? Same comment for ldaps test.

File: controllers/authrealm/suite_test.go [hunk 11]
```diff
@@ -631,16 +632,187 @@ var _ = Describe("Process AuthRealm LDAP: ", func() {
 			Expect(err).Should(BeNil())
 			Expect(len(dexServer.Spec.Connectors)).To(Equal(1))
 			Expect(dexServer.Spec.Connectors[0].LDAP.BindDN).To(Equal(BindDN))
+			Expect(dexServer.Spec.Connectors[0].Type).To(Equal(identitatemdexserverv1lapha1.ConnectorTypeLDAP))
+			Expect(dexServer.Spec.IngressCertificateRef.Name).To(Equal(authRealm.Spec.CertificatesSecretRef.Name))
+			Expect(dexServer.Spec.Connectors[0].LDAP.StartTLS).To(Equal(true))
```
Comment (robinbobbitt) [inline controllers/authrealm/suite_test.go:637]: Shall we check insecureNoSSL too? Same for ldaps test?

File: controllers/authrealm/dex.go (hunk 8 above)
Comment (leena-jawale) [reply controllers/authrealm/dex.go:447]: Yeah I agree with not allowing `insecure: true`. I will update the webhook check to not allow `insecure: true`

**Pull request merged**